### PR TITLE
Issue 15640 avoid duplicated messages

### PR DESCRIPTION
**Pull request by jdotcms**

### CODE REVIEW

**Review by wezell — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/api/system/event/message/SystemMessageEventUtil.java [hunk 1]
```diff
@@ -18,6 +18,51 @@
 public class SystemMessageEventUtil {
 
     private final SystemEventsAPI systemEventsAPI;
+    private static ThreadLocal<Map<Object, SystemEvent>> threadSystemMessageSet            = new ThreadLocal<>();
+    private static ThreadLocal<Boolean>                  accumulateUniqueMessagesPerThread = ThreadLocal.withInitial(()-> Boolean.FALSE);
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/api/system/event/message/SystemMessageEventUtil.java]: to be honest I didn't like the solution too, will work in another approach

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerUtil.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    } // getInstance.
+
+    private static final String NEW_LINE = Html.br() + Html.space(2);
+
+    /**
+     * Helper method to notify when container does not exist
+     * @param e {@link Exception}
+     * @param containerId {@link String}
+     */
+    public void notifyException(final Exception e, final String containerId) {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerUtil.java]: ok sounds good

File: dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java [hunk 3]
```diff
@@ -382,4 +384,36 @@ public static String getCurrentStackTraceAsString (final int limit) {
...
+
+        if (null != request) {
+
+            final PageMode pageMode = PageMode.get(request);
+            isPreviewOrEdit = (PageMode.EDIT_MODE == pageMode || PageMode.PREVIEW_MODE == pageMode);
+        } else {
+
+
+            for (final StackTraceElement stackTraceElement : e.getStackTrace()) {
+                if (stackTraceElement.getClassName().indexOf("EditMode") > -1 ||
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java]: It is old code, I have merge it with the PageMode when there is a http request available

File: dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java [hunk 4]
```diff
@@ -382,4 +384,36 @@ public static String getCurrentStackTraceAsString (final int limit) {
...
+     * @return boolean
+     */
+    public static boolean isPreviewOrEditMode(final Exception e, final HttpServletRequest request) {
+
+        boolean isPreviewOrEdit          = false;
+
+        if (null != request) {
+
+            final PageMode pageMode = PageMode.get(request);
+            isPreviewOrEdit = (PageMode.EDIT_MODE == pageMode || PageMode.PREVIEW_MODE == pageMode);
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java [hunk 5]
```diff
@@ -18,10 +18,7 @@
 import com.dotmarketing.exception.DotDataException;
 import com.dotmarketing.exception.DotRuntimeException;
 import com.dotmarketing.exception.DotSecurityException;
-import com.dotmarketing.portlets.containers.business.ContainerAPI;
-import com.dotmarketing.portlets.containers.business.ContainerFinderByIdOrPathStrategy;
-import com.dotmarketing.portlets.containers.business.LiveContainerFinderByIdOrPathStrategyResolver;
-import com.dotmarketing.portlets.containers.business.WorkingContainerFinderByIdOrPathStrategyResolver;
+import com.dotmarketing.portlets.containers.business.*;
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerUtil.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerUtil.java]: Done

File: dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java [hunk 6]
```diff
@@ -382,4 +384,36 @@ public static String getCurrentStackTraceAsString (final int limit) {
...
         return builder.toString();
     }
+
+    /**
+     * Determine if the exception happens on edit or preview mode, if the request is not null will use it to figure out
+     * @param e {@link Exception}
+     * @param request {@link HttpServletRequest}
+     * @return boolean
+     */
+    public static boolean isPreviewOrEditMode(final Exception e, final HttpServletRequest request) {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java]: Done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/ContainerLoader.java [hunk 7]
```diff
@@ -41,17 +45,27 @@ public InputStream writeObject(final VelocityResourceKey key)
...
+        try {
+            final Container container = strategy.isPresent() ?
+                    strategy.get().apply(key) : resolver.getDefaultStrategy().apply(key);
 
-        if(null == container) {
+            if (null == container) {
 
-            throw new DotStateException("cannot find container for : " +  key);
-        }
+                final DotStateException e = new DotStateException("Cannot find container for : " + key);
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/ContainerLoader.java]: Done

**Review by jgambarios — APPROVED** [review]

**Pull request merged**